activerecord/lib/active_record/migration.rb

```diff
@@ -935,8 +937,9 @@ def method_missing(method, *arguments, &block)
             end
           end
         end
-        return super unless connection.respond_to?(method)
-        connection.send(method, *arguments, &block)
+        execution_strategy = ActiveRecord.migration_strategy.new(connection, self)
```

Comment:
I think we'll need something along the lines of the memoization you had before 233cb000c7294a684fd1eca587e1ae902bbcaf9c.

If the trouble with memoizing was that `connection` can change, perhaps we start by removing `connection` from the strategy, and have it pull `@migration.connection` when it needs it? That doesn't feel like a long-term healthy arrangement, but I think my hope is that we're going to end up removing the existing connection-swapping in favour of strategy-swapping, so it should only be a passing requirement.